File: site/src/jsMain/kotlin/net/numa08/niconico_advertiser_list2/utils/VideoIdExtractor.kt
package net.numa08.niconico_advertiser_list2.utils

/**
 * ニコニコ動画のURLから動画IDを抽出するユーティリティ
 */
object VideoIdExtractor {
    /**
     * URLまたは動画IDから動画IDを抽出する
     * @param input ニコニコ動画のURLまたは動画ID
     * @return 動画ID（例: "sm12345678"）
     */
    fun extractVideoId(input: String): String? {
        // TODO: 実際の正規表現でURLから動画IDを抽出する実装に置き換える
        // 現時点ではモック実装として、入力をそのまま返す

        // 空白をトリミング
        val trimmed = input.trim()

        if (trimmed.isEmpty()) {
            return null
        }

        // モック実装: 入力がURL形式でない場合はそのまま動画IDとして扱う
        // 実装例:
        // - https://www.nicovideo.jp/watch/sm12345678 -> sm12345678
        // - https://nico.ms/sm12345678 -> sm12345678
        // - sm12345678 -> sm12345678

        return trimmed
    }
}

File: frontend/src/lib/videoIdExtractor.ts
// 動画ID抽出（仕様: FRONTEND_REQUIREMENTS.md FE-120〜FE-124）
// 現行KotlinのVideoIdExtractorを移植。既存ユニットテストのケース表が受け入れ基準

import { isNiconicoDomain, parseSimpleUrl, stripQueryAndFragment } from "./nicoUrl";

const WATCH_PATH_PREFIX = "/watch/";

/** `sm` + 1文字以上の数字のみを有効な動画IDとする（FE-121。大文字SMは不可） */
function isValidVideoId(candidate: string): boolean {
  return /^sm\d+$/.test(candidate);
}

/** URLまたは動画IDの入力から動画IDを抽出する。抽出できない場合はnull */
export function extractVideoId(input: string): string | null {
  const trimmed = input.trim();
  if (trimmed === "") {
    return null;
  }
  if (isValidVideoId(trimmed)) {
    return trimmed;
  }

  const parsed = parseSimpleUrl(trimmed);
  if (parsed === null || !isNiconicoDomain(parsed.host)) {
    return null;
  }
  if (!parsed.path.startsWith(WATCH_PATH_PREFIX)) {
    return null;
  }
  const candidate = stripQueryAndFragment(parsed.path.slice(WATCH_PATH_PREFIX.length));
  return isValidVideoId(candidate) ? candidate : null;
}
